File: tests/cpu/python/test_fft.py
"""FFT CPU binding tests."""
import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)


class TestFFTCPU:
    def test_fft(self):
        x_np = np.array([1, 2, 3, 4], dtype=np.float64)
        result = ins.fft(ins.from_numpy(x_np))
        expected = np.fft.fft(x_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_ifft(self):
        x_np = np.array([1, 2, 3, 4], dtype=np.float64)
        fft_np = np.fft.fft(x_np)
        result = ins.ifft(ins.from_numpy(fft_np.astype(np.complex128)))
        expected = np.fft.ifft(fft_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_fft_ifft_roundtrip(self):
        x_np = np.array([1, 2, 3, 4, 5, 6, 7, 8], dtype=np.float64)
        X = ins.fft(ins.from_numpy(x_np))
        x_back = ins.ifft(X)
        np.testing.assert_allclose(x_back.numpy().real, x_np, atol=1e-8)

    def test_rfft(self):
        x_np = np.array([1, 2, 3, 4], dtype=np.float64)
        result = ins.rfft(ins.from_numpy(x_np))
        expected = np.fft.rfft(x_np)
        assert result.numel() == expected.shape[0]

    def test_irfft(self):
        x_np = np.array([1, 2, 3, 4], dtype=np.float64)
        rfft_np = np.fft.rfft(x_np)
        result = ins.irfft(ins.from_numpy(rfft_np.astype(np.complex128)))
        expected = np.fft.irfft(rfft_np)
        assert result.numel() == expected.shape[0]

    def test_fftfreq(self):
        result = ins.fftfreq(8, 1.0)
        expected = np.fft.fftfreq(8, 1.0)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_rfftfreq(self):
        result = ins.rfftfreq(8, 1.0)
        expected = np.fft.rfftfreq(8, 1.0)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_fft2(self):
        x_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.fft2(ins.from_numpy(x_np))
        expected = np.fft.fft2(x_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_ifft2(self):
        x_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        fft_np = np.fft.fft2(x_np)
        result = ins.ifft2(ins.from_numpy(fft_np.astype(np.complex128)))
        expected = np.fft.ifft2(fft_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_fft_longer_signal(self):
        np.random.seed(42)
        x_np = np.random.randn(256)
        result = ins.fft(ins.from_numpy(x_np))
        expected = np.fft.fft(x_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-6)

    def test_fft_complex_input(self):
        x_np = np.array([1 + 1j, 2 + 2j, 3 + 3j, 4 + 4j], dtype=np.complex128)
        result = ins.fft(ins.from_numpy(x_np))
        expected = np.fft.fft(x_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_fft_ones(self):
        x_np = np.ones(8, dtype=np.float64)
        result = ins.fft(ins.from_numpy(x_np))
        expected = np.fft.fft(x_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_fft_power_of_two(self):
        x_np = np.random.randn(16)
        result = ins.fft(ins.from_numpy(x_np))
        expected = np.fft.fft(x_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_fft_non_power_of_two(self):
        x_np = np.random.randn(10)
        result = ins.fft(ins.from_numpy(x_np))
        expected = np.fft.fft(x_np)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)

    def test_fftfreq_custom_spacing(self):
        result = ins.fftfreq(16, 0.5)
        expected = np.fft.fftfreq(16, 0.5)
        np.testing.assert_allclose(result.numpy(), expected, atol=1e-8)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cpu/lua/test_fft.lua
-- FFT CPU binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu ~/.luarocks/bin/busted tests/cpu/lua/test_fft.lua

describe("FFT CPU Tests", function()
  local ins
  setup(function()
    ins = require("_insight")
    ins.set_device(ins.CPUPlace())
  end)

  it("fft 1d", function()
    local x = ins.from_table({ 1.0, 2.0, 3.0, 4.0 })
    local y = ins.fft(x)
    assert.is_not_nil(y)
    assert.are.equal(4, y.numel)
  end)

  it("ifft 1d", function()
    local x = ins.from_table({ 1.0, 2.0, 3.0, 4.0 })
    local y = ins.fft(x)
    local z = ins.ifft(y)
    assert.is_not_nil(z)
    assert.are.equal(4, z.numel)
  end)

  it("fft ifft roundtrip", function()
    local x = ins.from_table({ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 })
    local X = ins.fft(x)
    local x_back = ins.ifft(X)
    assert.is_not_nil(x_back)
    assert.are.equal(8, x_back.numel)
  end)

  it("rfft", function()
    local x = ins.from_table({ 1.0, 2.0, 3.0, 4.0 })
    local y = ins.rfft(x)
    assert.is_not_nil(y)
    assert.are.equal(3, y.numel)
  end)

  it("irfft", function()
    local x = ins.from_table({ 1.0, 2.0, 3.0, 4.0 })
    local y = ins.rfft(x)
    local z = ins.irfft(y)
    assert.is_not_nil(z)
    assert.are.equal(4, z.numel)
  end)

  it("fftfreq", function()
    local f = ins.fftfreq(8, 1.0)
    assert.is_not_nil(f)
    assert.are.equal(8, f.numel)
  end)

  it("rfftfreq", function()
    local f = ins.rfftfreq(8, 1.0)
    assert.is_not_nil(f)
    assert.are.equal(5, f.numel)
  end)

  it("rfft2", function()
    local x = ins.from_table({ { 1.0, 2.0 }, { 3.0, 4.0 } })
    local y = ins.rfft2(x)
    assert.is_not_nil(y)
  end)

  it("irfft2", function()
    local x = ins.from_table({ { 1.0, 2.0 }, { 3.0, 4.0 } })
    local y = ins.rfft2(x)
    local z = ins.irfft2(y)
    assert.is_not_nil(z)
    assert.are.equal(4, z.numel)
  end)

  it("fft longer signal", function()
    local t = {}
    for i = 1, 64 do
      t[i] = math.sin(2 * math.pi * i / 64)
    end
    local x = ins.from_table(t)
    local y = ins.fft(x)
    assert.is_not_nil(y)
    assert.are.equal(64, y.numel)
  end)

  it("fft ones", function()
    local t = {}
    for i = 1, 8 do
      t[i] = 1.0
    end
    local x = ins.from_table(t)
    local y = ins.fft(x)
    assert.is_not_nil(y)
    assert.are.equal(8, y.numel)
  end)

  it("fft non power of two", function()
    local t = {}
    for i = 1, 10 do
      t[i] = i * 1.0
    end
    local x = ins.from_table(t)
    local y = ins.fft(x)
    assert.is_not_nil(y)
    assert.are.equal(10, y.numel)
  end)

  it("fftshift", function()
    local x = ins.from_table({ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 })
    local y = ins.fftshift(x)
    assert.are.equal(8, y.numel)
    -- fftshift([1..8]) = [5,6,7,8,1,2,3,4]
    assert.are.equal(5.0, y[1])
    assert.are.equal(6.0, y[2])
    assert.are.equal(7.0, y[3])
    assert.are.equal(8.0, y[4])
    assert.are.equal(1.0, y[5])
    assert.are.equal(2.0, y[6])
    assert.are.equal(3.0, y[7])
    assert.are.equal(4.0, y[8])
  end)

  it("ifftshift", function()
    local x = ins.from_table({ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 })
    local y = ins.ifftshift(x)
    assert.are.equal(8, y.numel)
    -- ifftshift([1..8]) = [5,6,7,8,1,2,3,4]
    assert.are.equal(5.0, y[1])
    assert.are.equal(6.0, y[2])
    assert.are.equal(7.0, y[3])
    assert.are.equal(8.0, y[4])
    assert.are.equal(1.0, y[5])
    assert.are.equal(2.0, y[6])
    assert.are.equal(3.0, y[7])
    assert.are.equal(4.0, y[8])
  end)

  it("next_fast_len", function()
    local n = ins.next_fast_len(100)
    assert.is_true(n >= 100)
  end)

  it("fftfreq custom spacing", function()
    local f = ins.fftfreq(16, 0.5)
    assert.is_not_nil(f)
    assert.are.equal(16, f.numel)
  end)
end)
